core/src/replay_stage.rs

```diff
@@ -3060,43 +3065,54 @@ impl ReplayStage {
             match switch_fork_decision {
                 SwitchForkDecision::FailedSwitchThreshold(switch_proof_stake, total_stake) => {
                     let reset_bank = heaviest_bank_on_same_voted_fork;
-                    // If we can't switch and our last vote was on a non-duplicate/confirmed slot, then
-                    // reset to the the next votable bank on the same fork as our last vote,
-                    // but don't vote.
-
-                    // We don't just reset to the heaviest fork when switch threshold fails because
-                    // a situation like this can occur:
-
-                    /* Figure 1:
-                                  slot 0
-                                    |
-                                  slot 1
-                                /        \
-                    slot 2 (last vote)     |
-                                |      slot 8 (10%)
-                        slot 4 (9%)
-                    */
-
-                    // Imagine 90% of validators voted on slot 4, but only 9% landed. If everybody that fails
-                    // the switch threshold abandons slot 4 to build on slot 8 (because it's *currently* heavier),
-                    // then there will be no blocks to include the votes for slot 4, and the network halts
-                    // because 90% of validators can't vote
-                    info!(
-                        "Waiting to switch vote to {},
-                        resetting to slot {:?} for now,
-                        switch proof stake: {},
-                        threshold stake: {},
-                        total stake: {}",
-                        heaviest_bank.slot(),
-                        reset_bank.as_ref().map(|b| b.slot()),
-                        switch_proof_stake,
-                        total_stake as f64 * SWITCH_FORK_THRESHOLD,
-                        total_stake
-                    );
-                    failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
-                        heaviest_bank.slot(),
-                    ));
-                    reset_bank.map(|b| (b, switch_fork_decision))
+                    if last_vote_unable_to_land && reset_bank.is_some() {
+                        // If we can't switch, our last vote was on a non-duplicate/confirmed slot, but it is
```

Comment:
"our last vote was on a non-duplicate/confirmed slot" - why does this have to be true?

Reply:
If your last vote was on a duplicate, unconfirmed slot, you will get SwitchForkDecision::FailedSwitchDuplicateRollback, in this case I'm not sure we want to blindly stick to the current fork, because you want to avoid getting into a 50-50 split, so the best strategy might still be not voting for now.

Reply:
hmm ok, I see this is talking about the `FailedSwitchDuplicateRollback` but is inside the `FailedSwitchThreshold` case of the match statement which is kind of confusing.

A few things:
1.  Maybe this would be more appropriate as a comment on the `FailedSwitchDuplicateRollback` arm of the match statement
2. "If we can't switch, our last vote was on a non-duplicate/confirmed slot, but it is now outside slothash, it means we haven't voted for a while, and there was no hope", when we have a list of statements like this, it's clearer to number them in a list so that it's clear that these are assumptions and not that one implies the other. I was confused because I thought you were saying "If we can't switch," implies "our last vote was on a non-duplicate/confirmed slot,"

Reply:
ping on this comment

Reply:
Oops, somehow I missed last comment. Actually this is explaining the reasoning of choosing tip of the fork, I've rephrased the comments, hopefully this looks better?